lxd/db/storage_volume_snapshots.go

```diff
@@ -0,0 +1,147 @@
+// +build linux,cgo,!agent
+
+package db
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/lxc/lxd/lxd/db/query"
+	"github.com/lxc/lxd/shared"
+)
+
+// StoragePoolVolumSnapshoteCreate creates a new storage volume snapshot attached to a given
+// storage pool.
+func (c *Cluster) StoragePoolVolumeSnapshotCreate(project, volumeName, volumeDescription string, volumeType int, poolID int64, volumeConfig map[string]string, expiryDate time.Time) (int64, error) {
```

Comment:
@monstermunchkin this is quite a lot of code to mostly duplicate IMHO, and given the old function `StoragePoolVolumeCreate` was already "snapshot" aware, I think I'd be tempted to add the expiresAt (rather than expiryDate) argument to the old function. Either that or remove the snapshot awareness from the old function and then just have this function used for creating snapshot volumes. Rather than leave the ability to create snapshots still in the old function, albeit it without the expiry functionality.

Reply:
Yeah, this is something we need to clarify. I think it's cleaner to have a separate function which handles snapshots, instead of adding `expiryDate` to the "old" function. That's because I didn't want to pass `time.Time{}` in most of those calls as most of the them are used for non-snapshots.

I don't have a strong opinion on this, I just went for the IMHO cleaner solution.

Reply:
@monstermunchkin yes I see your point, makes sense. I don't mind either way, but if we do go with the new snapshot specific function, then I think it would be a good idea for future clarity to remove snapshot support from the old `StoragePoolVolumeCreate` function, and update any usage of it in a snapshot context to use the new function. Perhaps even add detection inside `StoragePoolVolumeCreate` to block creation of volumes with a snapshot name (similar to how we've done in backendLXD).